sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedSource.java

```diff
@@ -134,6 +136,106 @@ public Double getFractionConsumed() {
     }
 
     /**
+     * A constant to use as the return value for {@link #getParallelismConsumed()} or
+     * {@link #getParallelismRemaining()} when the exact value is unknown.
+     */
+    public static final long PARALLELISM_UNKNOWN = -1;
+
+    /**
+     * Returns the total amount of parallelism in the consumed (returned and processed) range of
+     * this reader's current {@link BoundedSource} (as would be returned by
+     * {@link #getCurrentSource}). This corresponds to all split point records (see
+     * {@link RangeTracker}) returned by this reader, <em>excluding</em> the current record.
+     *
+     * <p>Consider the following two examples. An input that can be read in parallel down to the
+     * individual records, such as {@link CountingSource#upTo}, is called "perfectly splittable".
+     * Conversely, an example of a non-perfectly splittable input is a block-compressed file format
+     * such as {@link AvroIO}, in which a block of records has to be read as a whole, but different
+     * blocks can be read in parallel.
+     *
+     * <ul>
+     * <li>Any {@link BoundedReader reader} that is unstarted (aka, has never had a call to
+     * {@link #start}) has a consumed parallelism of 0.
+     * <li>Any {@link BoundedReader reader} that has only returned its first element (aka,
+     * has never had a call to {@link #advance}) has a consumed parallelism of 0: the first element
+     * is the current element and is still being processed.
+     * <li>When processing record #30 (starting at 1) out of 50 in a perfectly splittable 50-record
+     * input, this value should be 29.
```

Comment:
Emphasize that this is because the 30th record hasn't been consumed yet, although that seems odd since wouldn't it mean that a split could take that record away? Or we'll always split and leave at least one record on the current worker?